File: backend/src/services/ShareService.ts
import { prisma } from '../db/prisma';
import { ShareType } from '@prisma/client';
import { generateUID } from './UIDService';
import { hashPassword } from './PasswordService';
import * as StorageService from './StorageService';

// ─── Types ──────────────────────────────────────────────

export interface CreateFileShareInput {
  files: Array<{
    filename: string;
    buffer: Buffer;
    mimeType: string;
    size: number;
  }>;
  expiresIn?: string; // '1h', '6h', '24h', '7d', '30d'
  password?: string;
}

export interface CreateTextShareInput {
  title?: string;
  content: string;
  language?: string;
  expiresIn?: string;
  password?: string;
}

export interface ShareMetadata {
  uid: string;
  type: ShareType;
  isPrivate: boolean;
  expiresAt: string | null;
  createdAt: string;
  totalSize: string;
  fileCount?: number;
  files?: Array<{
    id: string;
    filename: string;
    mimeType: string;
    size: string;
  }>;
  textShare?: {
    title: string | null;
    language: string;
    // content is NOT included in metadata — requires auth for private shares
  } | null;
}

// ─── Helpers ────────────────────────────────────────────

const EXPIRY_MAP: Record<string, number> = {
  '1h': 60 * 60 * 1000,
  '6h': 6 * 60 * 60 * 1000,
  '24h': 24 * 60 * 60 * 1000,
  '7d': 7 * 24 * 60 * 60 * 1000,
  '30d': 30 * 24 * 60 * 60 * 1000,
};

function calculateExpiry(expiresIn?: string): Date | null {
  if (!expiresIn || !EXPIRY_MAP[expiresIn]) return null;
  return new Date(Date.now() + EXPIRY_MAP[expiresIn]);
}

// ─── Service ────────────────────────────────────────────

/**
 * Create a file share: upload files to R2, create DB records, return UID.
 */
export async function createFileShare(input: CreateFileShareInput): Promise<string> {
  const uid = await generateUID();
  const expiresAt = calculateExpiry(input.expiresIn);
  const isPrivate = !!input.password;
  const passwordHash = input.password ? await hashPassword(input.password) : null;
  const totalSize = input.files.reduce((sum, f) => sum + f.size, 0);

  // Upload all files to R2
  const fileRecords = await Promise.all(
    input.files.map(async (file) => {
      const storageKey = StorageService.generateStorageKey(uid, file.filename);
      await StorageService.uploadFile(storageKey, file.buffer, file.mimeType);
      return {
        filename: file.filename,
        storageKey,
        mimeType: file.mimeType,
        size: BigInt(file.size),
      };
    })
  );

  // Create share + file records in a transaction
  await prisma.share.create({
    data: {
      uid,
      type: ShareType.FILE,
      isPrivate,
      passwordHash,
      expiresAt,
      totalSize: BigInt(totalSize),
      files: {
        create: fileRecords,
      },
    },
  });

  return uid;
}

/**
 * Create a text share: store text content in DB, return UID.
 */
export async function createTextShare(input: CreateTextShareInput): Promise<string> {
  const uid = await generateUID();
  const expiresAt = calculateExpiry(input.expiresIn);
  const isPrivate = !!input.password;
  const passwordHash = input.password ? await hashPassword(input.password) : null;
  const contentSize = Buffer.byteLength(input.content, 'utf-8');

  await prisma.share.create({
    data: {
      uid,
      type: ShareType.TEXT,
      isPrivate,
      passwordHash,
      expiresAt,
      totalSize: BigInt(contentSize),
      textShare: {
        create: {
          title: input.title || null,
          content: input.content,
          language: input.language || 'plaintext',
        },
      },
    },
  });

  return uid;
}

/**
 * Get share metadata by UID. Does NOT include text content.
 */
export async function getShareByUID(uid: string): Promise<ShareMetadata | null> {
  const share = await prisma.share.findUnique({
    where: { uid },
    include: {
      files: {
        select: {
          id: true,
          filename: true,
          mimeType: true,
          size: true,
        },
      },
      textShare: {
        select: {
          title: true,
          language: true,
        },
      },
    },
  });

  if (!share) return null;

  // Check if expired
  if (share.expiresAt && share.expiresAt < new Date()) {
    return null;
  }

  return {
    uid: share.uid,
    type: share.type,
    isPrivate: share.isPrivate,
    expiresAt: share.expiresAt?.toISOString() || null,
    createdAt: share.createdAt.toISOString(),
    totalSize: share.totalSize.toString(),
    fileCount: share.files.length,
    files: share.files.map((f) => ({
      id: f.id,
      filename: f.filename,
      mimeType: f.mimeType,
      size: f.size.toString(),
    })),
    textShare: share.textShare,
  };
}

/**
 * Get text content for a share (used after auth verification for private shares).
 */
export async function getTextContent(uid: string): Promise<{ title: string | null; content: string; language: string } | null> {
  const share = await prisma.share.findUnique({
    where: { uid },
    include: { textShare: true },
  });

  if (!share || !share.textShare) return null;
  if (share.expiresAt && share.expiresAt < new Date()) return null;

  return {
    title: share.textShare.title,
    content: share.textShare.content,
    language: share.textShare.language,
  };
}

/**
 * Get the password hash for a share (used in verification flow).
 */
export async function getSharePasswordHash(uid: string): Promise<string | null> {
  const share = await prisma.share.findUnique({
    where: { uid },
    select: { passwordHash: true },
  });
  return share?.passwordHash || null;
}

/**
 * Get file details for download (including internal storageKey).
 */
export async function getFileForDownload(fileId: string): Promise<{ storageKey: string; filename: string; shareUid: string; isPrivate: boolean } | null> {
  const file = await prisma.file.findUnique({
    where: { id: fileId },
    include: {
      share: {
        select: { uid: true, isPrivate: true, expiresAt: true },
      },
    },
  });

  if (!file) return null;
  if (file.share.expiresAt && file.share.expiresAt < new Date()) return null;

  return {
    storageKey: file.storageKey,
    filename: file.filename,
    shareUid: file.share.uid,
    isPrivate: file.share.isPrivate,
  };
}

/**
 * Get all files for a share (used for ZIP download).
 */
export async function getShareFiles(uid: string): Promise<Array<{ storageKey: string; filename: string }> | null> {
  const share = await prisma.share.findUnique({
    where: { uid },
    include: { files: { select: { storageKey: true, filename: true } } },
  });

  if (!share) return null;
  if (share.expiresAt && share.expiresAt < new Date()) return null;

  return share.files;
}

File: backend/app/services/share_service.py
"""
share_service.py — Core CRUD operations for shares.
Replaces: backend/src/services/ShareService.ts
"""

from datetime import datetime, timedelta, timezone
from uuid import uuid4

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models import Share, File, TextShare, ShareType
from app.services.uid_service import generate_uid, format_uid
from app.services.password_service import hash_password
from app.services import storage_service

# ─── Expiry mapping ──────────────────────────────────────

EXPIRY_MAP: dict[str, timedelta] = {
    "30m": timedelta(minutes=30),
    "1h": timedelta(hours=1),
    "6h": timedelta(hours=6),
    "24h": timedelta(hours=24),
    "7d": timedelta(days=7),
    "30d": timedelta(days=30),
}


def _calculate_expiry(expires_in: str | None) -> datetime | None:
    if not expires_in or expires_in not in EXPIRY_MAP:
        return None
    return datetime.now(timezone.utc) + EXPIRY_MAP[expires_in]


# ─── Create file share ──────────────────────────────────


async def create_file_share(
    db: AsyncSession,
    files: list[dict],
    expires_in: str | None = None,
    password: str | None = None,
) -> str:
    """Create a file share: upload files to storage, create DB records, return UID."""
    uid = await generate_uid(db)
    expires_at = _calculate_expiry(expires_in)
    is_private = bool(password)
    pw_hash = await hash_password(password) if password else None
    total_size = sum(f["size"] for f in files)

    # Upload all files to storage
    file_records = []
    for f in files:
        storage_key = storage_service.generate_storage_key(uid, f["filename"])
        await storage_service.upload_file(storage_key, f["buffer"], f["mime_type"])
        file_records.append(
            File(
                id=str(uuid4()),
                filename=f["filename"],
                storage_key=storage_key,
                mime_type=f["mime_type"],
                size=f["size"],
            )
        )

    share = Share(
        id=str(uuid4()),
        uid=uid,
        type=ShareType.FILE,
        is_private=is_private,
        password_hash=pw_hash,
        expires_at=expires_at,
        total_size=total_size,
        files=file_records,
    )

    db.add(share)
    await db.commit()
    return uid


# ─── Create text share ──────────────────────────────────


async def create_text_share(
    db: AsyncSession,
    content: str,
    title: str | None = None,
    language: str | None = None,
    expires_in: str | None = None,
    password: str | None = None,
) -> str:
    """Create a text share: store in DB, return UID."""
    uid = await generate_uid(db)
    expires_at = _calculate_expiry(expires_in)
    is_private = bool(password)
    pw_hash = await hash_password(password) if password else None
    content_size = len(content.encode("utf-8"))

    text_share = TextShare(
        id=str(uuid4()),
        title=title,
        content=content,
        language=language or "plaintext",
    )

    share = Share(
        id=str(uuid4()),
        uid=uid,
        type=ShareType.TEXT,
        is_private=is_private,
        password_hash=pw_hash,
        expires_at=expires_at,
        total_size=content_size,
        text_share=text_share,
    )

    db.add(share)
    await db.commit()
    return uid


# ─── Get share metadata ─────────────────────────────────


async def get_share_by_uid(db: AsyncSession, uid: str) -> dict | None:
    """Get share metadata by UID. Does NOT include text content."""
    result = await db.execute(
        select(Share)
        .where(Share.uid == uid)
        .options(selectinload(Share.files), selectinload(Share.text_share))
    )
    share = result.scalar_one_or_none()
    if not share:
        return None

    # Check expiry
    if share.expires_at and share.expires_at < datetime.now(timezone.utc):
        return None

    return {
        "uid": share.uid,
        "type": share.type.value,
        "isPrivate": share.is_private,
        "expiresAt": share.expires_at.isoformat() if share.expires_at else None,
        "createdAt": share.created_at.isoformat(),
        "totalSize": str(share.total_size),
        "fileCount": len(share.files),
        "files": [
            {
                "id": f.id,
                "filename": f.filename,
                "mimeType": f.mime_type,
                "size": str(f.size),
            }
            for f in share.files
        ],
        "textShare": (
            {"title": share.text_share.title, "language": share.text_share.language}
            if share.text_share
            else None
        ),
    }


# ─── Get text content ───────────────────────────────────


async def get_text_content(db: AsyncSession, uid: str) -> dict | None:
    """Get text content for a share."""
    result = await db.execute(
        select(Share).where(Share.uid == uid).options(selectinload(Share.text_share))
    )
    share = result.scalar_one_or_none()
    if not share or not share.text_share:
        return None
    if share.expires_at and share.expires_at < datetime.now(timezone.utc):
        return None
    return {
        "title": share.text_share.title,
        "content": share.text_share.content,
        "language": share.text_share.language,
    }


# ─── Get password hash ──────────────────────────────────


async def get_share_password_hash(db: AsyncSession, uid: str) -> str | None:
    """Get the password hash for a share."""
    result = await db.execute(select(Share.password_hash).where(Share.uid == uid))
    row = result.scalar_one_or_none()
    return row


# ─── Get file for download ──────────────────────────────


async def get_file_for_download(db: AsyncSession, file_id: str) -> dict | None:
    """Get file details for download (including internal storageKey)."""
    result = await db.execute(
        select(File).where(File.id == file_id).options(selectinload(File.share))
    )
    file = result.scalar_one_or_none()
    if not file:
        return None
    if file.share.expires_at and file.share.expires_at < datetime.now(timezone.utc):
        return None
    return {
        "storage_key": file.storage_key,
        "filename": file.filename,
        "mime_type": file.mime_type,
        "share_uid": file.share.uid,
        "is_private": file.share.is_private,
    }


# ─── Get all files for ZIP ──────────────────────────────


async def get_share_files(db: AsyncSession, uid: str) -> list[dict] | None:
    """Get all files for a share (used for ZIP download)."""
    result = await db.execute(
        select(Share).where(Share.uid == uid).options(selectinload(Share.files))
    )
    share = result.scalar_one_or_none()
    if not share:
        return None
    if share.expires_at and share.expires_at < datetime.now(timezone.utc):
        return None
    return [{"storage_key": f.storage_key, "filename": f.filename} for f in share.files]
